### PR TITLE
Add overflow_closures and overflow_match_arms opts

### PR DESCRIPTION
**Pull request by spinda**
cc https://github.com/rust-lang-nursery/rustfmt/issues/1791

## `overflow_closures`

Allow overflow on closures

- **Default value**: `true`
- **Possible values**: `true`, `false`

#### `true`:

```rust
result.and_then(|maybe_value| match maybe_value {
    None => ...,
    Some(value) => ...,
})
```

#### `false`:

```rust
result.and_then(|maybe_value| {
    match maybe_value {
        None => ...,
        Some(value) => ...,
    }
})
```

## `overflow_match_arms`

Allow overflow on match arms

- **Default value**: `true`
- **Possible values**: `true`, `false`

#### `true`:

```rust
match lorem {
    None => if ipsum {
        println!("Hello World");
    },
    Some(dolor) => ...,
}
```

#### `false`:

```rust
match lorem {
    None => {
        if ipsum {
            println!("Hello World");
        }
    }
    Some(dolor) => ...,
}
```

### CODE REVIEW

**Comment by topecongiro** [comment]
Thank you for you PR! Could you please add tests for `overflow_closures` as well?

**Comment by spinda** [comment]
Oops, I thought I'd done that. Will add.

File: src/expr.rs [hunk 1]
```diff
@@ -688,6 +688,13 @@ fn rewrite_closure_expr(
     if classify::expr_requires_semi_to_be_stmt(left_most_sub_expr(expr)) {
         rewrite = and_one_line(rewrite);
     }
+    rewrite = rewrite.and_then(|rw| {
+        if !context.config.overflow_closures() && rw.contains('\n') {
```
Comment (emilio) [inline src/expr.rs]: (Totally not a `rustfmt` expert, so could totally be wrong, but won't this undo any other formatting on everything else inside the expression?)

Comment (spinda) [inline src/expr.rs]: Look at how `rewrite_closure_expr` is used:

https://github.com/rust-lang-nursery/rustfmt/blob/6fb17c41301e5a98748882cc6ad6a627def8eb80/src/expr.rs#L629-L648

https://github.com/rust-lang-nursery/rustfmt/blob/6fb17c41301e5a98748882cc6ad6a627def8eb80/src/expr.rs#L650-L653

It's called to attempt to format the closure body expression without wrapping it in a block. If it fails and returns `None`, rustfmt will fall back to `rewrite_closure_block` and format the expression with a block surrounding it. `rewrite_closure_block` will perform any formatting that's still appropriate for the expression in that context.

Comment (emilio) [inline src/expr.rs]: Yup, ok, fair enough. That happens when drive-by reviewing, thanks for explaining :)

**Comment by nrc** [comment]
Bikeshedding, I would prefer `closures_force_block` rather than `overflow_closures` (and likewise for match arms). "Overflow" suggests something being too full, which is not the case here, we're just using a block (you'd need to flip the true/false too).

**Comment by spinda** [comment]
@nrc That would seem to imply always forcing a block for closures and match arms, whereas this option only forces a block when the respective constructs "overflow" a single line.

**Comment by nrc** [comment]
Perhaps `multline_closure_forces_block` or something similar? The issues with the current names are that they don't indicate what happens on overflow, and "overflow" is not a term we use in Rustfmt at the moment.

**Comment by spinda** [comment]
I like `multiline_closure_forces_block`. My original choice of `overflow` did in fact come from the rustfmt code, though; see https://github.com/rust-lang-nursery/rustfmt/pull/1898/files#diff-180bb3ad456cc10517d878fb7795341cR1708 for example (`can_be_overflowed`).

**Comment by nrc** [comment]
Thanks for making the changes!
